File: app/src/main/java/com/mintthursday/StepsAdapter.java
package com.mintthursday;

import android.graphics.Color;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class StepsAdapter extends RecyclerView.Adapter<StepsAdapter.StepsViewHolder> implements RecyclerRowMoveCallback.RecyclerRowMoveTouchHelper {
    private List<Step> stepsList = new ArrayList<>();

    public void setItems(List<Step> steps) {
        stepsList = steps;
        notifyDataSetChanged();
    }

    public void addItem(Step step) {
        stepsList.add(step);
        notifyDataSetChanged();
    }

    public void clearItems() {
        stepsList.clear();
        notifyDataSetChanged();
    }
    public List<Step> getItems(){
        return stepsList;
    }


    @NonNull
    @NotNull
    @Override
    public StepsAdapter.StepsViewHolder onCreateViewHolder(@NonNull @NotNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.view_step, parent, false);
        return new StepsViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull @NotNull StepsAdapter.StepsViewHolder holder, int position) {
        holder.bind(stepsList.get(position));
        holder.stepInstr.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
            }

            @Override
            public void afterTextChanged(Editable s) {
                int i = holder.getAdapterPosition();
                stepsList.get(i).setStepInstruction(s.toString().trim());
            }
        });

    }

    @Override
    public int getItemCount() {
        return stepsList.size();
    }

    @Override
    public void onRowMoved(int from, int to) {
        if (from < to) {
            for (int i = from; i < to; i++) {
                Collections.swap(stepsList, i, i + 1);
            }
        } else {
            for (int i = from; i > to; i--) {
                Collections.swap(stepsList, i, i - 1);
            }
        }
        notifyItemMoved(from, to);
    }

    @Override
    public void onRowSelected(StepsViewHolder stepsViewHolder) {
        stepsViewHolder.itemView.setBackgroundColor(Color.GRAY);

    }

    @Override
    public void onClearRow(StepsViewHolder stepsViewHolder) {
        stepsViewHolder.itemView.setBackgroundColor(Color.WHITE);

    }

    class StepsViewHolder extends RecyclerView.ViewHolder {
        private TextView stepInstr;
        private ImageView handle;

        public StepsViewHolder(View itemView) {
            super(itemView);
            stepInstr = itemView.findViewById(R.id.step);
            handle = itemView.findViewById(R.id.handle);
        }


        public void bind(Step step) {
            stepInstr.setText(step.getStepInstruction());

        }

    }
}

File: app/src/main/java/com/mintthursday/recipe/creation/stepsadapter/StepsAdapter.kt
package com.mintthursday.recipe.creation.stepsadapter

import android.graphics.Color
import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageButton
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.mintthursday.R
import com.mintthursday.models.Step
import com.mintthursday.recipe.creation.stepsadapter.RecyclerRowMoveCallback.RecyclerRowMoveTouchHelper
import com.mintthursday.recipe.creation.stepsadapter.StepsAdapter.StepsViewHolder
import java.util.*

class StepsAdapter : RecyclerView.Adapter<StepsViewHolder>(), RecyclerRowMoveTouchHelper {

    var items: MutableList<Step> = mutableListOf()
        private set

    fun setItems(items: List<Step>) {
        this.items = items.toMutableList()
        notifyItemRangeInserted(0, items.size)
    }

    fun addItem(step: Step) {
        items.add(step)
        notifyItemInserted(items.size)
    }

    fun deleteItem(position: Int) {
        items.removeAt(position)
        notifyItemRemoved(position)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): StepsViewHolder {
        val view = LayoutInflater.from(parent.context)
                .inflate(R.layout.view_step, parent, false)
        return StepsViewHolder(view)
    }

    override fun onBindViewHolder(holder: StepsViewHolder, position: Int) {
        holder.bind(items[position])

    }

    override fun getItemCount(): Int {
        return items.size
    }

    override fun onRowMoved(from: Int, to: Int) {
        if (from < to) {
            for (i in from until to) {
                Collections.swap(items, i, i + 1)
            }
        } else {
            for (i in from downTo to + 1) {
                Collections.swap(items, i, i - 1)
            }
        }
        notifyItemMoved(from, to)
    }

    override fun onRowSelected(stepsViewHolder: StepsViewHolder?) {
        stepsViewHolder?.itemView?.setBackgroundColor(Color.GRAY)
    }

    override fun onClearRow(stepsViewHolder: StepsViewHolder?) {
        stepsViewHolder?.itemView?.setBackgroundColor(Color.WHITE)
    }

    inner class StepsViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        private val stepInstr: TextView = itemView.findViewById(R.id.step)
        private val handle: ImageView = itemView.findViewById(R.id.handle)
        private val btnStepRemove: ImageButton = itemView.findViewById(R.id.step_remove)

        fun bind(step: Step) {
            stepInstr.text = step.stepInstruction
            stepInstr.addTextChangedListener(object : TextWatcher {
                override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
                override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {}
                override fun afterTextChanged(s: Editable) {
                    items[adapterPosition].stepInstruction = s.toString().trim { it <= ' ' }
                }
            })
        }

        init {
            btnStepRemove.setOnClickListener() {
                deleteItem(adapterPosition)
            }
        }

    }
}